microsoft/azurelinux
# [mic][iso] generate PXE-bootable ISO images.

###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*

---

###### Summary 
This change enables the Azures Linux Image Customizer to generate liveos ISO images that are 'PXE-ready'. To be 'PXE-ready', the user should be able to take the generated artifacts, copy them to a PXE server, and is then able to boot PXE clients - without any modifications to any of the artifacts.

To enable such experience, the work involves:
1. Ensure that Dracut's livenet module (which supports downloading liveos ISOs) works correctly.
2. Include the livenet module in the MIC generated liveos ISO images.
3. Generate a grub.cfg with kernel parameters that correctly configures Dracut's livenet module.
4. Generate a folder with OS artifacts that are directly copy-able to the PXE server for deployment.

The first requirement is broken in Dracut 102 and earlier versions.  The fix is covered by this [PR: Enable Dracut's livenet rootfs handling when systemd-networkd is in use.](https://github.com/microsoft/azurelinux/pull/10706).

The other three requirements are implemented in this PR.

The changes include:
- interface:
  - configuration: new configuration to specify the download url for the iso.
  - command-line: new parameter to export the iso image contents into a folder for easier deployment unto a pxe server.
- implementation:
  - implemented the above interface changes.
  - extended the set of parameters that must be saved between runs to include the new pxe parameters.
  - updated the code to emit a pxe grub.cfg and a pxe dracut config.

###### Regression Risk

Small.

This is a new code path enabling a new flow. However, there's some overlap with saving MIC internal state between runs.

###### Change Log  



- See commits.

###### Does this affect the toolchain?  



- no

###### Associated issues  


- 9739

###### Links to CVEs  
- n/a

###### Test Methodology

Unit Tests
- Ran MIC liveos ISO tests (tests include coverage for new scenario).

Functional Tests
- Deploy a PXE server on the network.
- Use MIC to generate a liveos iso (vhdx->iso) and specify the new `--output-pxe-artifacts-dir` parameter.
  - Attach the iso to a hyper-v VM
    - Verify the VM boots, and the user can log-in.
  - Deploy the iso artifacts to a PXE server and boot a PXE client on the same network.
    - Verify the PXE client boots, and that the user can log-in.
- Use MIC to generate a liveos iso (iso->iso) and specify the new `--output-pxe-artifacts-dir` parameter.
  - Attach the iso to a hyper-v VM
    - Verify the VM boots, and the user can log-in.
  - Deploy the iso artifacts to a PXE server and boot a PXE client on the same network.
    - Verify the PXE client boots, and that the user can log-in.
    - Verify user settings from the first run (vhdx-iso) have been retained.
